libdnf/comps/environment/environment.cpp

```diff
@@ -0,0 +1,317 @@
+    for (EnvironmentId environment_id : environment_ids) {
+        Solvable * solvable = pool.id2solvable(environment_id.id);
+        if (const char * translation = solvable_lookup_str_poollang(solvable, SOLVABLE_DESCRIPTION)) {
+            // Return translation only if it's different from the untranslated string
+            // (solvable_lookup_str_lang returns the untranslated string if there is no translation).
+            const char * untranslated = solvable_lookup_str(solvable, SOLVABLE_DESCRIPTION);
+            if (translation != untranslated && strcmp(translation, untranslated) != 0) {
+                return std::string(translation);
+            }
+        }
+    }
+    // If no translation was found, return the untranslated string.
+    return this->get_description();
+}
+
+
+std::string Environment::get_order() const {
+    libdnf::solv::Pool & pool = get_pool(base);
+    return lookup_str(pool, environment_ids, SOLVABLE_ORDER);
+}
+
+std::vector<std::string> Environment::get_groups() {
+    // Return groups if they are already loaded
+    if (!groups.empty()) {
+        return groups;
+    }
+
+    libdnf::solv::Pool & pool = get_pool(base);
+
+    // Use only the first (highest priority) solvable for group lists
+    Solvable * solvable = pool.id2solvable(environment_ids[0].id);
+    std::string group_solvable_name;
+
+    if (solvable->dep_requires) {
+        for (Id * r_id = solvable->repo->idarraydata + solvable->dep_requires; *r_id; ++r_id) {
+            std::string group_solvable_name = pool.id2str(*r_id);
+            groups.push_back(group_solvable_name.substr(group_solvable_name.find(":") + 1, std::string::npos));
+        }
+    }
+    return groups;
+}
+
+std::vector<std::string> Environment::get_optional_groups() {
+    // Return optional_groups if they are already loaded
+    if (!optional_groups.empty()) {
+        return optional_groups;
+    }
+
+    libdnf::solv::Pool & pool = get_pool(base);
+
+    // Use only the first (highest priority) solvable for group lists
+    Solvable * solvable = pool.id2solvable(environment_ids[0].id);
+    std::string group_solvable_name;
+
+    if (solvable->dep_suggests) {
+        for (Id * r_id = solvable->repo->idarraydata + solvable->dep_suggests; *r_id; ++r_id) {
+            std::string group_solvable_name = pool.id2str(*r_id);
+            optional_groups.push_back(group_solvable_name.substr(group_solvable_name.find(":") + 1, std::string::npos));
+        }
+    }
+    return optional_groups;
+}
+
+std::set<std::string> Environment::get_repos() const {
+    libdnf::solv::Pool & pool = get_pool(base);
+
+    std::set<std::string> result;
+    for (EnvironmentId environment_id : environment_ids) {
+        Solvable * solvable = pool.id2solvable(environment_id.id);
+        result.emplace(solvable->repo->name);
+    }
+    return result;
+}
+
+
+bool Environment::get_installed() const {
+    auto repos = get_repos();
+    return repos.find("@System") != repos.end();
```

Comment:
In most cases I agree, but here it doesn't seem better, because I'd have to call get_repos() twice.
`return get_repos().find("@System") != get_repos().end();`

Reply:
Oh, you're right in this case, of course.